File: js-packages/misc/src/utils/crypto.ts
import * as Interfaces from '../interfaces'
import {
    unserializeToArrayBuffer,
    splitFirstOnly,
    b64tobuffer,
    utf8encoder,
    utf8decoder,
} from './encoding'
import { MaybePromise } from '../typing'
import {
    mapDeriveAlgorithms,
    mapEncryptionAlgorithms,
    mapSignatureAlgorithms,
    ParamsType,
    KeyType,
    SignatureAlgorithm,
    EncryptionAlgorithm,
} from './base_crypto'

export {
    mapDeriveAlgorithms,
    mapEncryptionAlgorithms,
    mapSignatureAlgorithms,
    DEFAULT_SIGNATURE_ALGORITHM,
    DEFAULT_ASYMMETRIC_ENCRYPTION_ALGORITHM,
    DEFAULT_SYMMETRIC_ENCRYPTION_ALGORITHM,
    DEFAULT_DERIVE_ALGORITHM,
    KeyType,
} from './base_crypto'

export {
    KeyTypeError,
    UnknownAlgorithm,
    EmptyKeyError,
} from './base_crypto_legacy'

import { UnknownAlgorithm } from './base_crypto_legacy'
export type FullKeyType = KeyType | Interfaces.RawInput

export function findWorkingAlgorithms(
    algorithms: string[],
    domain:
        | 'hash'
        | 'derive'
        | 'symmetric'
        | 'asymmetric'
        | 'signature'
        | 'all'
): string[] {
    // only signature needs no extra set
    // js sets are insertion order stable
    const algos: Set<string> = new Set()
    for (const algo of algorithms) {
        let found = null
        if (
            (domain == 'all' || domain == 'hash') &&
            mapDeriveAlgorithms[algo]?.type == 'hash'
        ) {
            found = mapDeriveAlgorithms[algo].serializedName
        } else if (
            (domain == 'all' || domain == 'derive') &&
            mapDeriveAlgorithms[algo]?.type == 'derive'
        ) {
            found = mapDeriveAlgorithms[algo].serializedName
        } else if (
            (domain == 'all' || domain == 'asymmetric') &&
            mapEncryptionAlgorithms[algo]?.type == 'asymmetric'
        ) {
            found = mapEncryptionAlgorithms[algo].serializedName
        } else if (
            (domain == 'all' || domain == 'symmetric') &&
            mapEncryptionAlgorithms[algo]?.type == 'symmetric'
        ) {
            found = mapEncryptionAlgorithms[algo].serializedName
        } else if (
            (domain == 'all' || domain == 'signature') &&
            mapSignatureAlgorithms[algo]
        ) {
            found = mapSignatureAlgorithms[algo].serializedName
        }
        if (found && !algos.has(found)) {
            algos.add(found)
        }
    }
    return [...algos]
}

async function key_helper(key: MaybePromise<FullKeyType>) {
    const key_cleaned = await key
    if (key_cleaned instanceof CryptoKey) {
        return key_cleaned
    }
    return await unserializeToArrayBuffer(key_cleaned)
}

export async function derive(
    inp: MaybePromise<string | ArrayBuffer>,
    { params, algorithm }: { params?: any; algorithm: string } = {
        algorithm: '',
    }
): Promise<{
    data: ArrayBuffer
    params: any
    serializedName: string
    serialized: string
}> {
    let inp_cleaned: ArrayBuffer | string = await inp
    if (!algorithm && typeof inp_cleaned == 'string') {
        const splitted = splitFirstOnly(inp_cleaned)
        algorithm = splitted[0]
        inp_cleaned = splitted[1]
    }

    const entry = mapDeriveAlgorithms['' + algorithm]
    if (!entry) {
        throw new UnknownAlgorithm('invalid algorithm: ' + algorithm)
    }
    if (typeof inp_cleaned == 'string') {
        const result = await entry.deserialize(inp_cleaned, params)
        inp_cleaned = result.data
        params = result.params
    }
    const result = await entry.derive(inp_cleaned, params)
    return {
        ...result,
        serializedName: entry.serializedName,
        serialized: `${entry.serializedName}:${await entry.serialize(result)}`,
    }
}

export async function deserializeDerivedString(
    inp: MaybePromise<string>,
    { params, algorithm }: { params?: any; algorithm: string } = {
        algorithm: '',
    } // params are here fallback?
) {
    let inp_cleaned: string = await inp
    if (!algorithm) {
        const splitted = splitFirstOnly(inp_cleaned)
        algorithm = splitted[0]
        inp_cleaned = splitted[1]
    }
    const entry = mapDeriveAlgorithms['' + algorithm]
    if (!entry) {
        throw Error('invalid algorithm: ' + algorithm)
    }
    const result = await entry.deserialize(inp_cleaned, params)
    return {
        ...result,
        serializedName: entry.serializedName,
        serialized: `${entry.serializedName}:${inp_cleaned}`,
    }
}

export async function deriveString(
    data: MaybePromise<string | ArrayBuffer>,
    options: { params?: any; algorithm: string } = { algorithm: '' }
): Promise<string> {
    return (await derive(data, options)).serialized
}

export async function serializeDerive(
    inp: MaybePromise<{
        data: ArrayBuffer
        params: any
        serializedName: string
    }>
): Promise<string> {
    const result = await inp
    const entry = mapDeriveAlgorithms[result.serializedName]
    return `${entry.serializedName}:${await entry.serialize(result)}`
}

export async function generateEncryptionKey(
    { params, algorithm }: { params?: any; algorithm: string } = {
        algorithm: '',
    }
): Promise<{
    params?: any
    serializedName: string
    key: ArrayBuffer
}> {
    const entry = mapEncryptionAlgorithms['' + algorithm]
    if (!entry) {
        throw Error('invalid algorithm: ' + algorithm)
    }
    return {
        ...(await entry.generateKey(params)),
        serializedName: entry.serializedName,
    }
}

export async function generateSignKey(
    { params, algorithm }: { params?: any; algorithm: string } = {
        algorithm: '',
    }
): Promise<{
    params?: any
    serializedName: string
    key: ArrayBuffer
}> {
    const entry = mapSignatureAlgorithms['' + algorithm]
    if (!entry) {
        throw Error('invalid algorithm: ' + algorithm)
    }
    return {
        ...(await entry.generateKey(params)),
        serializedName: entry.serializedName,
    }
}

export async function encrypt(
    key: MaybePromise<FullKeyType>,
    data: MaybePromise<string | ArrayBuffer>,
    { params, algorithm }: { params?: ParamsType; algorithm: string } = {
        algorithm: '',
    }
): Promise<{
    data: ArrayBuffer
    params: any
    serializedName: string
    key: ArrayBuffer
}> {
    let data_cleaned = await unserializeToArrayBuffer(data)
    const entry = mapEncryptionAlgorithms['' + algorithm]
    if (!entry) {
        throw new UnknownAlgorithm('invalid algorithm: ' + algorithm)
    }
    return {
        ...(await entry.encrypt(await key_helper(key), data_cleaned, params)),
        serializedName: entry.serializedName,
    }
}

export async function encryptString(
    key: MaybePromise<FullKeyType>,
    data: MaybePromise<string | ArrayBuffer>,
    options: { params?: ParamsType; algorithm: string }
): Promise<string> {
    const result = await encrypt(key, data, options)
    const entry = mapEncryptionAlgorithms[result.serializedName]
    if (!entry) {
        throw new UnknownAlgorithm(
            'invalid algorithm: ' + result.serializedName
        )
    }
    let serializedParams = await entry.serializeParams(result.params)
    if (serializedParams) {
        serializedParams = `${serializedParams}:`
    }
    return `${entry.serializedName}:${serializedParams}${Buffer.from(
        result.data
    ).toString('base64')}`
}

export async function serializeEncryptionParams({
    params,
    serializedName,
}: {
    params: any
    serializedName: string
}) {
    const entry = mapEncryptionAlgorithms[serializedName]
    if (!entry) {
        throw Error('invalid algorithm: ' + serializedName)
    }
    if (entry.serializeParams) {
        return `${serializedName}:${await entry.serializeParams(params)}`
    } else {
        return `${serializedName}:`
    }
}

export async function decrypt(
    key: MaybePromise<FullKeyType>,
    data: MaybePromise<string | ArrayBuffer>,
    { params, algorithm }: { params: ParamsType | string; algorithm?: string }
): Promise<{
    data: ArrayBuffer
    params: any
    serializedName: string
    key: ArrayBuffer
}> {
    if (!algorithm && typeof params == 'string') {
        const splitted = splitFirstOnly(params)
        algorithm = splitted[0]
        const entry = mapEncryptionAlgorithms['' + algorithm]
        params = (await entry.deserialize(splitted[1])).params
    }
    const entry = mapEncryptionAlgorithms['' + algorithm]
    if (!entry) {
        throw new UnknownAlgorithm('invalid algorithm: ' + algorithm)
    }
    const data_cleaned = await unserializeToArrayBuffer(data)
    if (data_cleaned === undefined) {
        throw new Error('Empty data')
    }
    try {
        return {
            ...(await entry.decrypt(
                await key_helper(key),
                data_cleaned,
                params
            )),
            serializedName: entry.serializedName,
        }
    } catch (exc) {
        // console.error('decrypt failed:', key_cleaned, data_cleaned, params)

        throw exc
    }
}
export async function decryptString(
    key: MaybePromise<FullKeyType>,
    data: MaybePromise<string>,
    {
        params,
        algorithm,
    }: {
        params?: any
        algorithm?: string
    } = {}
): Promise<{
    data: ArrayBuffer
    params: any
    serializedName: string
    key: ArrayBuffer
}> {
    let data_cleaned = await data
    if (!algorithm) {
        const splitted = splitFirstOnly(data_cleaned)
        algorithm = splitted[0]
        data_cleaned = splitted[1]
    }

    const entry = mapEncryptionAlgorithms['' + algorithm]
    if (!entry) {
        throw new UnknownAlgorithm('invalid algorithm: ' + algorithm)
    }
    const result = await entry.deserialize(data_cleaned, params)
    params = result.params
    if (!result.data) {
        throw Error('no data to decrypt provided')
    }
    return await decrypt(await key, result.data, {
        params,
        algorithm: entry.serializedName,
    })
}
export async function sign(
    key: MaybePromise<FullKeyType>,
    data: MaybePromise<string | ArrayBuffer>,
    { algorithm }: { algorithm: string }
): Promise<string> {
    const entry = mapSignatureAlgorithms['' + algorithm]
    if (!entry) {
        throw Error('invalid algorithm: ' + algorithm)
    }
    const data_cleaned = await unserializeToArrayBuffer(data)
    try {
        return `${entry.serializedName}:${await entry.sign(
            await key_helper(key),
            data_cleaned
        )}`
    } catch (exc) {
        console.error('sign parameters:', algorithm, data_cleaned)
        throw exc
    }
}
export async function verify(
    key: MaybePromise<FullKeyType>,
    signature: MaybePromise<string>,
    data: MaybePromise<string | ArrayBuffer>,
    { algorithm }: { algorithm?: string } = {}
): Promise<boolean> {
    let signature_cleaned = await signature
    if (!algorithm) {
        const splitted = splitFirstOnly(signature_cleaned)
        algorithm = splitted[0]
        signature_cleaned = splitted[1]
    }

    const entry = mapSignatureAlgorithms['' + algorithm]
    if (!entry) {
        throw new UnknownAlgorithm('invalid algorithm: ' + algorithm)
    }
    const data_cleaned = await unserializeToArrayBuffer(data)
    return entry.verify(await key_helper(key), signature_cleaned, data_cleaned)
}

export async function toHashableKey(
    key: MaybePromise<FullKeyType>,
    {
        algorithm,
        sign,
        raiseOnSymmetric = false,
    }: { algorithm?: string; sign?: boolean; raiseOnSymmetric?: boolean } = {}
): Promise<{ serializedName: string; key: ArrayBuffer; params: ParamsType }> {
    let keyEntry: undefined | SignatureAlgorithm | EncryptionAlgorithm =
        undefined
    if (sign === undefined || sign) {
        keyEntry = mapSignatureAlgorithms['' + algorithm]
    }
    if (!keyEntry && !sign) {
        keyEntry = mapEncryptionAlgorithms['' + algorithm]
        if (raiseOnSymmetric && keyEntry.type == 'symmetric') {
            throw new Error('not an asymmetric algorithm')
        }
    }
    if (!keyEntry) {
        throw new UnknownAlgorithm('invalid key algorithm: ' + algorithm)
    }
    const hashableKey = await keyEntry.toHashableKey(await key_helper(key))
    return {
        serializedName: keyEntry.serializedName,
        key: hashableKey,
        params: {},
    }
}

export async function toPublicKey(
    key: MaybePromise<FullKeyType>,
    options: { algorithm?: string; sign?: boolean } = {}
) {
    return await toHashableKey(key, { ...options, raiseOnSymmetric: true })
}

export async function hashKey(
    key: MaybePromise<FullKeyType>,
    {
        keyAlgorithm,
        sign,
        deriveAlgorithm,
        deriveParams,
    }: {
        keyAlgorithm: string
        sign?: boolean
        deriveAlgorithm: string
        deriveParams?: any
    }
): Promise<{
    data: ArrayBuffer
    params: any
    serializedName: string
    key: ArrayBuffer
    serialized: string
}> {
    const hashableKey = (
        await toHashableKey(key, { sign, algorithm: keyAlgorithm })
    ).key
    const result = await derive(hashableKey, {
        algorithm: deriveAlgorithm,
        params: deriveParams,
    })
    return {
        ...result,
        key: hashableKey,
    }
}

export async function hashKeyString(
    key: MaybePromise<FullKeyType>,
    options: {
        keyAlgorithm: string
        deriveAlgorithm: string
        sign?: boolean
        deriveParams?: any
    }
): Promise<string> {
    const result = await hashKey(key, options)
    return result.serialized
}

export async function buildKeyHashSignature(
    key: MaybePromise<FullKeyType>,
    data: MaybePromise<ArrayBuffer>,
    {
        keyAlgorithm,
        deriveAlgorithm,
    }: {
        keyAlgorithm: string
        deriveAlgorithm: string
    }
): Promise<string> {
    const signaturePromise = sign(key, data, { algorithm: keyAlgorithm })
    const result = await hashKey(key, {
        sign: true,
        keyAlgorithm: keyAlgorithm,
        deriveAlgorithm: deriveAlgorithm,
    })
    return `${Buffer.from(utf8encoder.encode(result.serialized)).toString(
        'base64'
    )}:${await signaturePromise}`
}

export function splitKeyHashSignature(data: string): [string, string] {
    const splitted = splitFirstOnly(data)
    if (!splitted[0]) {
        throw Error('not a keyHashSignature string')
    }
    return [utf8decoder.decode(b64tobuffer(splitted[0])), splitted[1]]
}

File: secretgraph/core/utils/crypto.py
from base64 import b64decode, b64encode
from collections.abc import Awaitable
from dataclasses import dataclass
from inspect import isawaitable
from logging import getLogger
from typing import Iterable, Literal, TypeVar

from ..exceptions import UnknownAlgorithm
from .base_crypto import (
    CryptoResult,
    DeriveResult,
    DeserializeResult,
    KeyResult,
    KeyType,
    ParamsType,
    mapDeriveAlgorithms,
    mapEncryptionAlgorithms,
    mapSignatureAlgorithms,
)

logger = getLogger(__name__)

DataInputType = TypeVar(
    "DataInputType", str, bytes, memoryview, Awaitable[str | bytes | memoryview]
)
KeyInputType = TypeVar(
    "KeyInputType", str, bytes, memoryview, Awaitable[str | bytes | memoryview]
)
ParamsInputType = TypeVar(
    "ParamsInputType",
    ParamsType,
    None,
)
ParamsInputType2 = TypeVar(
    "ParamsInputType2",
    ParamsInputType,
    str,
)


@dataclass(frozen=True, kw_only=True)
class FullDeriveResult(DeriveResult):
    serializedName: str
    serialized: str


@dataclass(frozen=True, kw_only=True)
class FullCryptoResult(CryptoResult):
    serializedName: str


@dataclass(frozen=True, kw_only=True)
class FullDeserializeResult(DeserializeResult):
    serializedName: str


@dataclass(frozen=True, kw_only=True)
class FullKeyResult(KeyResult):
    serializedName: str


@dataclass(frozen=True, kw_only=True)
class FullHashKeyResult(FullDeriveResult):
    key: bytes


async def data_helper(data: DataInputType) -> bytes | str:
    if isawaitable(data):
        data = await data
    if isinstance(data, memoryview):
        data = data.tobytes()
    return data


def findWorkingAlgorithms(
    algorithms: Iterable[str],
    domain: Literal["hash"]
    | Literal["derive"]
    | Literal["symmetric"]
    | Literal["asymmetric"]
    | Literal["signature"]
    | Literal["all"],
    failhard=False,
) -> list[str]:
    # only dicts are insertion order stable
    algos = {}
    for algo in algorithms:
        found = None
        if (
            (domain == "all" or domain == "hash")
            and algo in mapDeriveAlgorithms
            and mapDeriveAlgorithms[algo].type == "hash"
        ):
            found = mapDeriveAlgorithms[algo].serializedName
        elif (
            (domain == "all" or domain == "derive")
            and algo in mapDeriveAlgorithms
            and mapDeriveAlgorithms[algo].type == "derive"
        ):
            found = mapDeriveAlgorithms[algo].serializedName
        elif (
            (domain == "all" or domain == "asymmetric")
            and algo in mapEncryptionAlgorithms
            and mapEncryptionAlgorithms[algo].type == "asymmetric"
        ):
            found = mapEncryptionAlgorithms[algo].serializedName
        elif (
            (domain == "all" or domain == "symmetric")
            and algo in mapEncryptionAlgorithms
            and mapEncryptionAlgorithms[algo].type == "symmetric"
        ):
            found = mapEncryptionAlgorithms[algo].serializedName
        elif (domain == "all" or domain == "signature") and mapSignatureAlgorithms[
            algo
        ]:
            found = mapSignatureAlgorithms[algo].serializedName
        elif failhard:
            raise UnknownAlgorithm("Unknown algorithm: " + algo)
        if found and found not in algos:
            algos[found] = True
    return list(algos.keys())


async def derive(
    data: DataInputType, params: ParamsInputType = None, algorithm: str = ""
) -> FullDeriveResult:
    data = await data_helper(data)
    if not algorithm and isinstance(data, str):
        splitted = data.split(":", 1)
        algorithm = splitted[0]
        data = splitted[1]

    entry = mapDeriveAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)

    if isinstance(data, str):
        result = await entry.deserialize(data, params)
        data = result.data
        params = result.params
    assert isinstance(data, (bytes, Iterable))
    result = await entry.derive(data, params)
    return FullDeriveResult(
        data=result.data,
        params=result.params,
        serializedName=entry.serializedName,
        serialized=f"{entry.serializedName}:{await entry.serialize(result)}",
    )


async def deserializeDerivedString(
    inp: Awaitable[str] | str, params: ParamsInputType = None, algorithm: str = ""
) -> FullDeriveResult:
    if isawaitable(inp):
        inp = await inp
    if not algorithm:
        splitted = inp.split(":", 1)
        algorithm = splitted[0]
        inp = splitted[1]
    entry = mapDeriveAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("unknown algorithm")

    result = await entry.deserialize(inp, params)
    return FullDeriveResult(
        params=result.params,
        data=result.data,
        serializedName=entry.serializedName,
    )


async def deriveString(
    data: DataInputType, params: ParamsInputType = None, algorithm: str = ""
) -> str:
    result = await derive(data, params=params, algorithm=algorithm)
    return result.serialized


async def serializeDerive(inp: FullDeriveResult | Awaitable[FullDeriveResult]) -> str:
    if isawaitable(inp):
        inp = await inp
    entry = mapDeriveAlgorithms[inp.serializedName]
    return f"{entry.serializedName}:{await entry.serialize(inp)}"


async def generateEncryptionKey(
    algorithm: str = "", params: ParamsInputType = None
) -> FullKeyResult:
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    if isawaitable(params):
        params = await params
    result = await entry.generateKey(params)
    return FullKeyResult(
        key=result.key, params=result.params, serializedName=entry.serializedName
    )


async def generateSignKey(
    algorithm: str = "", params: ParamsInputType = None
) -> FullKeyResult:
    entry = mapSignatureAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    if isawaitable(params):
        params = await params
    result = await entry.generateKey(params)
    return FullKeyResult(
        key=result.key, params=result.params, serializedName=entry.serializedName
    )


async def encrypt(
    key: KeyInputType,
    data: DataInputType,
    params: ParamsInputType = None,
    algorithm: str = "",
) -> FullCryptoResult:
    data = await data_helper(data)
    if isinstance(data, str):
        data = b64decode(data)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    assert isinstance(data, bytes)
    result = await entry.encrypt(key, data, params)
    return FullCryptoResult(
        data=result.data,
        key=result.key,
        params=result.params,
        serializedName=entry.serializedName,
    )


async def encryptString(
    key: KeyboardInterrupt,
    data: DataInputType,
    params: ParamsInputType = None,
    algorithm: str = "",
) -> str:
    result = await encrypt(key, data, params=params, algorithm=algorithm)
    entry = mapEncryptionAlgorithms[result.serializedName]
    serializedParams = await entry.serializeParams(result.params)
    if serializedParams:
        serializedParams = f"{serializedParams}:"
    return f"{entry.serializedName}:{serializedParams}{b64encode(result.data).decode()}"


async def serializeEncryptionParams(params: ParamsInputType, algorithm: str) -> str:
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    if entry.serializeParams:
        return f"{entry.serializedName}:{await entry.serializeParams(params)}"
    else:
        return f"{entry.serializedName}:"


async def deserializeEncryptedString(
    data: Awaitable[str] | str, params: ParamsInputType = None, algorithm: str = ""
) -> FullDeserializeResult:
    if isawaitable(data):
        data = await data
    if not algorithm:
        splitted = data.split(":", 1)
        algorithm = splitted[0]
        data = splitted[1]
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("unknown algorithm")

    result = await entry.deserialize(data, params)
    return FullDeserializeResult(
        params=result.params,
        data=result.data,
        serializedName=entry.serializedName,
    )


async def decrypt(
    key: KeyInputType,
    data: DataInputType,
    algorithm: str = "",
    params: ParamsInputType2 = None,
) -> FullCryptoResult:
    if not algorithm and isinstance(params, str):
        result = await deserializeEncryptedString(params)
        algorithm = result.serializedName
        params = result.params
    data = await data_helper(data)
    if isinstance(data, str):
        data = b64decode(data)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    result = await entry.decrypt(key, data, params)
    return FullCryptoResult(
        data=result.data,
        params=result.params,
        key=result.key,
        serializedName=entry.serializedName,
    )


async def getDecryptor(
    key: KeyInputType,
    algorithm: str = "",
    params: ParamsInputType2 = None,
):
    if not algorithm and isinstance(params, str):
        result = await deserializeEncryptedString(params)
        algorithm = result.serializedName
        params = result.params
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)
    return await entry.decryptor(key, params)


async def getEncryptor(
    key: KeyInputType,
    algorithm: str = "",
    params: ParamsInputType2 = None,
) -> FullCryptoResult:
    if not algorithm and isinstance(params, str):
        result = await deserializeEncryptedString(params)
        algorithm = result.serializedName
        params = result.params
    entry = mapEncryptionAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)
    return await entry.encryptor(key, params)


async def decryptString(
    key: KeyInputType,
    data: str | Awaitable[str],
    params: ParamsInputType = None,
    algorithm: str = "",
) -> FullCryptoResult:
    result = await deserializeEncryptedString(data, params=params, algorithm=algorithm)
    return await decrypt(key, data, algorithm=result.serializedName, params=params)


async def sign(
    key: KeyInputType, data: DataInputType, algorithm: str, prehashed=False
) -> str:
    entry = mapSignatureAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    data = await data_helper(data)
    if isinstance(data, str):
        data = b64decode(data)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)

    try:
        return f"{entry.serializedName}:${await entry.sign(key, data, prehashed=prehashed)}"
    except Exception as exc:
        logger.error("sign parameters: %s, %s", algorithm, key)
        raise exc


async def getSignatureHasher(algorithm: str):
    entry = mapSignatureAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    return entry.getHasher()


async def verify(
    key: KeyType,
    signature: str | Awaitable[str],
    data: DataInputType,
    algorithm: str = "",
    prehashed=False,
) -> bool:
    if isawaitable(signature):
        signature = await signature
    if not algorithm:
        splitted = signature.split(":", 1)
        algorithm = splitted[0]
        signature = splitted[1]
    entry = mapSignatureAlgorithms.get(algorithm)
    if not entry:
        raise UnknownAlgorithm("invalid algorithm: " + algorithm)
    data = await data_helper(data)
    if isinstance(data, str):
        data = b64decode(data)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)

    return entry.verify(key, signature, data, prehashed=prehashed)


async def toHashableKey(
    key: KeyType | Awaitable[KeyType],
    algorithm: str,
    sign: bool | None = None,
    raiseOnSymmetric=False,
) -> FullKeyResult:
    keyEntry = None
    if sign is None or sign:
        keyEntry = mapSignatureAlgorithms.get(algorithm)
    if not keyEntry and not sign:
        keyEntry = mapEncryptionAlgorithms.get(algorithm)
        if keyEntry.type == "symmetric":
            raise ValueError("not an asymmetric algorithm")
    if not keyEntry:
        raise UnknownAlgorithm("invalid key algorithm: " + algorithm)
    key = await data_helper(key)
    if isinstance(key, str):
        key = b64decode(key)
    hashableKey = await keyEntry.toHashableKey(key)
    return FullKeyResult(key=hashableKey, serializedName=keyEntry.serializedName)


async def toPublicKey(
    key: KeyType | Awaitable[KeyType],
    algorithm: str,
    sign: bool | None = None,
) -> FullKeyResult:
    return await toHashableKey(
        key, algorithm=algorithm, sign=sign, raiseOnSymmetric=True
    )


async def hashKey(
    key: KeyType | Awaitable[KeyType],
    keyAlgorithm: str,
    deriveAlgorithm: str,
    sign: bool | None = None,
    deriveParams: ParamsType = None,
) -> FullHashKeyResult:
    result1 = await toHashableKey(key, algorithm=keyAlgorithm, sign=sign)
    result2 = await derive(
        result1.key,
        algorithm=deriveAlgorithm,
        params=deriveParams,
    )
    return FullHashKeyResult(
        data=result2.data,
        params=result2.params,
        serializedName=result2.serializedName,
        serialized=result2.serialized,
        key=result1.key,
    )
